### DISCUSSION TITLE
Dependencies

### DISCUSSION THREAD

**Discussion by viperML** [nvim-neorocks/lz.n#47] [Q&A] [answered]
How should a user express the dependencies between two plugins?

Let's say we have two plugins A and B:

- A is only explicitely required for ACommand
- B is only required for filetypes `.b`, but requires A to be loaded before

I believe we can use `require('lz.n').trigger_load("a")` in the `before` function of B?

**Comment by mrcjkb** [comment]
Hey 👋 

If a dependency is simply a Lua library (that doesn't source any plugin scripts), then I would recommend just to add it to the runtimepath (or the `package.path`).
It won't have any noticeable impact on startup time. 

Otherwise, you can use the `before` hook. 
Although I consider designing plugins in a way that users need to worry about loading/configuration order an anti pattern.

**Reply by viperML** [reply]
>  Although I consider designing plugins in a way that users need to worry about loading/configuration order an anti pattern.

The issue I am running is rather on the Lua side, as `opt` packages are not added to package.path until they are `:packadd`'d.

For example, take this dependency between nvim-autopairs and nvim-cmp:

```lua
      require("lz.n").load {
        {
          "nvim-autopairs",
          after = function()
            -- configure autopairs
          end
        },
        {
          "nvim-cmp",
          before = function()
            require("lz.n").trigger_load("nvim-autopairs")
          end,
          after = function()
            -- configure nvim-cmp, nvim-autopairs has been packadd'ed and configured
          end
        }
      }
```

Running the `trigger_load` fails with `Package nvim-autopairs not found`. Moreover, after peeking in `trigger_load`, it seems that is not idempotent, as the `after` function will get re-run for each `trigger_load` (perhaps it is not suited for this?)

**Reply by viperML** [reply]
I might as well add my whole `opt` to the `package.path` (why doesn't neovim only add `start`?)

**Reply by mrcjkb** [reply]
> Running the trigger_load fails with Package nvim-autopairs not found.

Is the error message "Package ... not found" or "Plugin... not found"?
If it's the latter, you may be running into the same problem as before, with the plugin not being registered properly, or being removed from the internal state.

**Reply by mrcjkb** [reply]
> Moreover, after peeking in trigger_load, it seems that is not idempotent, as the after function will get re-run for each trigger_load

It was intended to be used by handlers, which are expected to make sure they don't call it more than once for the same plugin.

**Reply by viperML** [reply]
> Is the error message "Package ... not found" or "Plugin... not found"?

Oops, it's `Plugin` not found. `:lua =require("lz.n.state")` shows `nvim-autopairs` in the plugin list, while if I call `vim.notify(vim.inspect("lz.state"))` in the `before` of `cmp` , the plugin list is empty.

```lua
      require("lz.n").load {
        {
          "nvim-autopairs",
          after = function()
            -- configure autopairs
            vim.notify("nvim-autopairs loaded")
          end,
        },
        {
          "nvim-cmp",
          before = function()
            vim.notify(vim.inspect(require("lz.n.state"))) --> plugins = {}
            require("lz.n").trigger_load("nvim-autopairs")
          end,
          after = function()
            -- configure nvim-cmp, nvim-autopairs has been packadd'ed and configured
            vim.notify("nvim-cmp loaded")
          end
        }
      }
```

**Reply by viperML** [reply]
In any case, if we forget about `trigger_load`, then I have no way of making sure `nvim-autopairs` gets loaded before `nvim-cmp` ....

**Reply by mrcjkb** [reply]
> if I call vim.notify(vim.inspect("lz.state")) in the before of cmp , the plugin list is empty.

That would indicate that `load` is being called with an empty list.

**Reply by BirdeeHub** [reply]
So, you should be able to call trigger_load actually.

handlers already have to deal with add and del being called for plugins they dont have in their list.

This would be no different, and would in fact ENSURE they are only called once, because trigger_load would then remove it from all the other handlers.

**Reply by BirdeeHub** [reply]
however in the snippet you showed, I dont see any conditions? So if the plugins look like [that](https://github.com/nvim-neorocks/lz.n/discussions/47#discussioncomment-10358848) in your config, they are actually being loaded eagerly? And thus trigger_load might not do anything because idk if they are added to state if they are eager.

EDIT: startup plugins (plugins that you didnt give a condition/event) are loaded before anything is added to state. So if they are both being loaded eagerly, neither of them would be in state yet.

Also you forgot the .n in require('lz.n.state') in your inspect statement, although im assuming you typed it right in the editor

**Reply by viperML** [reply]
> EDIT: startup plugins (plugins that you didnt give a condition/event) are loaded before anything is added to state. So if they are both being loaded eagerly, neither of them would be in state yet.

So my question is more in terms of the correcteness of ordering. Right now I have to assume that lz walks the list of start plugins, and loads them in sequential order, which means that order matters. It is not clear what would be the ordering if I set both to an event, such as DeferredUIEnter.

**Reply by viperML** [reply]
> trigger_load would then remove it from all the other handlers.

I can't follow this. On my experimentation, I can call `trigger_load` for a `cmd`-dependant (or any other non-start I guess?) plugin multiple times, and lz will call `after` every time, not once.

**Reply by BirdeeHub** [reply]
yes you could still accidentally call trigger_load many times and it would call the plugin's hooks each time.

But if you called it 1 time, none of the handlers would call it again automatically. So as long as you make sure YOU only call it once, you should be fine.

As far as what order they would happen in DeferredUIEnter, unfortunately I dont know.

**Reply by BirdeeHub** [reply]
That being said, I solved this stuff differently

[I literally just throw all the ones that should load together into the same spec and remap the load function](https://github.com/BirdeeHub/birdeeSystems/blob/b7fc5aec31a835512672cfaa7fe01a2196fbd938/common/birdeevim/lua/birdee/plugins/telescope.lua#L45-L53)

Its a bit messy and manual, but it does work!

**Reply by viperML** [reply]
> yes you could still accidentally call trigger_load many times and it would call the plugin's hooks each time.

What if I have two plugins B and C, that depend on A? Using `trigger_load` on the after/before of B and C is not correct, as you might trigger A twice.

And to be clear, I'm not saying `trigger_load` is bad, but rather that it is a low-level tool and we don't have some idempotent alternative.

**Reply by BirdeeHub** [reply]
Well, usually you dont actually have this issue in practice, because most plugins that multiple other ones depend on tend to be as close to pure lua as possible? Or at least make use of autoload? That means it is safe to load them at startup because they wont run into issues anyway.

But you could always hit it with the old vim.g.pluginA_loaded = 1 from the way vim plugins do it?

**Reply by BirdeeHub** [reply]
Unsure but maybe I can write a handler to help and post it here. Unsure tho

**Reply by BirdeeHub** [reply]
I also might make trigger load ignore multiple calls if I can. Working on it in was_loaded branch of my fork

**Reply by mrcjkb** [reply]
~I think making `trigger_load` idempotent would be a good idea.~

~Created an issue: #51~

**Reply by mrcjkb** [reply]
I've put some more thought into this. See https://github.com/nvim-neorocks/lz.n/issues/51#issuecomment-2294890207.

`trigger_load` should not be called from within a plugin spec's hook (I'll update the docs on that).

> What if I have two plugins B and C, that depend on A? Using trigger_load on the after/before of B and C is not correct, as you might trigger A twice.

@viperML do you have a concrete example for this that cannot be solved by bundling configurations of plugins and their extensions?

**Reply by viperML** [reply]
> @viperML do you have a concrete example for this that cannot be solved by bundling configurations of plugins and their extensions?

Admittedly, all the situations I've run into, the dependency doesn't have any `/plugin` folder, so I can just use `start` instead of `opt+packadd`. For my case, these dependencies were orgmode+org-bullets, dap+dap-python and telescope+telescope-fzf-native.

The only sitation I've run into for now is nvim-paredit. I want to be on `ft = {"scheme"}`, but it does a `require` on `nvim-tresitter`, which I have on `event = "DeferredUIEnter"`. When opening a scheme file, nvim-paredit triggers before the DeferredUIEnter, causing a require error. I may be able to work around this by adding `nvim-treesitter` to my `package.path`.



-- Note that while I've had a neovim configuration for years, only this month I've started to move to it as my main editor, so my opinions might not reflect a user with a bit more vision of the ecosystem.

**Reply by mrcjkb** [reply]
Hmm, nvim-paredit should stop depending on nvim-treesitter and use Neovim's built-in API.
The nvim-treesitter Lua API [is deprecated and will be removed](https://github.com/nvim-treesitter/nvim-treesitter/issues/4767).

The maintainers also recommend against lazy-loading nvim-treesitter.
But yes, in the meantime, you should be able to work around that.

**Reply by viperML** [reply]
> The maintainers also recommend against lazy-loading nvim-treesitter

That I didn't know!

**Comment by mrcjkb** [comment, marked as answer]
I've added [a section to the readme](https://github.com/nvim-neorocks/lz.n?tab=readme-ov-file#plugin-dependencies) on plugin dependencies.

**Comment by BirdeeHub** [comment]
A custom handler that loads a plugin right before load of another plugin

For the reasons above, this is rarely useful, but it CAN exist, here it is.

Useage: 

Put the below handler in a file in your config.

At the start of your config run:

`require('lze').register_handler(require("that.file"))`

In a spec:

`dep_of = { "plugin", "names", "that", "depend", "on", "the", "spec", "this", "is", "in" }`

```lua
local trigger_load = require("lz.n").trigger_load
local states = require("lz.n.handler.state").new()
---@type table<string, true>
local called = {}

---@type lz.n.Handler
---@diagnostic disable-next-line: missing-fields
local M = {
    spec_field = "dep_of",
    lookup = states.lookup_plugin
}

---@param plugin lz.n.Plugin
function M.add(plugin)
    local dep_of = plugin.dep_of

    ---@type string[]
    local needed_by = {}
    if type(dep_of) == "table" then
        ---@cast dep_of string[]
        needed_by = dep_of
    elseif type(dep_of) == "string" then
        needed_by = { dep_of }
    else
        return
    end
    for _, name in ipairs(needed_by) do
        if called[name] == true then
            trigger_load(plugin)
            return
        end
    end
    for _, name in ipairs(needed_by) do
        states.insert(name, plugin)
    end
end

---@param pname string
function M.del(pname)
    states.del(pname)
    called[pname] = true
    if states.has_pending_plugins(pname) then
        states.each_pending(pname,
            function (p)
                states.del(p.name)
                trigger_load(p)
            end
        )
    end
end

return M
```

**Reply by BirdeeHub** [reply]
Is this handler suitable for the wiki?

**Reply by mrcjkb** [reply]
sure, why not? 😁

**Reply by BirdeeHub** [reply]
:)

**Reply by BirdeeHub** [reply]
Whenever I figure out how to PR to it I'll add it haha I dont know how, its not its own repo and Ive never used a github wiki before XD

**Reply by mrcjkb** [reply]
you should be able to just edit it.

**Reply by BirdeeHub** [reply]
I dont see an edit button anywhere XD

and I can clone the repo but it has no repo page so I cant fork it? If I can from the cli I failed to do so because I couldnt figure it out in lazygit. And my fork that my PRs have been from doesnt have a wiki

**Reply by BirdeeHub** [reply]
https://github.com/nvim-neorocks/lz.n/wiki/Custom-handler-examples#dependency-handler

Added it :)

**Reply by mrcjkb** [reply]
thanks. 
One minor thing: I would suggest renaming the `dep_of` field to something like `load_before`.
"dep_of" implies that a plugin is a dependency of other plugins. luasnip isn't really a dependency of nvim-cmp.
